### PR TITLE
Add service_account.Credentials.with_claims

### LINKED ISSUE #129: Can't update additional claims with service account credentials

### PR DESCRIPTION
**Pull request by jonparrott**
Resolves #129.

### CODE REVIEW

File: google/auth/jwt.py [hunk 1]
```diff
@@ -426,13 +426,15 @@ def with_claims(self, issuer=None, subject=None, audience=None,
         Returns:
             google.auth.jwt.Credentials: A new credentials instance.
         """
+        new_additional_claims = self._additional_claims.copy()
```
Comment (dhermes) [inline google/auth/jwt.py]: You worried about nesting? `deepcopy`?

File: google/auth/jwt.py [hunk 2]
```diff
@@ -426,13 +426,15 @@ def with_claims(self, issuer=None, subject=None, audience=None,
...
+        new_additional_claims.update(additional_claims or {})
```
Comment (dhermes) [inline google/auth/jwt.py]: If you made `additional_claims=()` by default you could ditch the `or {}` (FWIW `()` is not strictly needed, any "empty" iterable will do)

File: google/oauth2/service_account.py [hunk 3]
```diff
@@ -246,6 +246,29 @@ def with_subject(self, subject):
             subject=subject,
             additional_claims=self._additional_claims.copy())
 
+    def with_claims(self, additional_claims):
```
Comment (dhermes) [inline google/oauth2/service_account.py]: This is mostly / totally copied from the `jwt` module?

Comment (jonparrott) [inline google/oauth2/service_account.py]: Yep.

File: google/auth/jwt.py (hunk 2 above)
Comment (jonparrott) [inline google/auth/jwt.py]: Eh, that borders on too clever for my likes.

File: google/auth/jwt.py (hunk 1 above)
Comment (jonparrott) [inline google/auth/jwt.py]: Nah, not horribly concerned about deep nesting, it's pretty rare to do that in the jwt payload.

Comment (jonparrott) [inline google/auth/jwt.py]: But if you want to be extra sure I can do deepcopy.

File: google/auth/jwt.py (hunk 2 above)
Comment (dhermes) [inline google/auth/jwt.py]: That would be the least clever thing I've heard called too clever.

File: google/oauth2/service_account.py (hunk 3 above)
Comment (dhermes) [inline google/oauth2/service_account.py]: Possibly to DRY?

File: google/auth/jwt.py (hunk 1 above)
Comment (dhermes) [inline google/auth/jwt.py]: Up to you

File: google/oauth2/service_account.py (hunk 3 above)
Comment (jonparrott) [inline google/oauth2/service_account.py]: Nah, jwt accepts a different set of arguments here (like audience, etc.). One one hand I know there's some duplicating here, but I'm not sure if there's a good way to refactor without messiness.

File: google/auth/jwt.py (hunk 2 above)
Comment (jonparrott) [inline google/auth/jwt.py]: 😆

**Comment by dhermes** [comment]
LGTM I suppose. It seems you want to eat the teeny cost of `or {}` every time, so it's all good?

**Comment by jonparrott** [comment]
>It seems you want to eat the teeny cost of or {} every time, so it's all good?

The cost of that vs. the cost of copy/deepcopy is insignificant.

I'm going to update this to use deepcopy, but I'm finishing something else up first. Thanks for the review!

**Pull request merged**